### PR TITLE
Fixed #37084 -- Added CSP nonce context processor system check.

### PR DESCRIPTION
**Pull request by MiladZarour**
#### Trac ticket number

ticket-37084

#### Branch description

Added a security system check warning when `ContentSecurityPolicyMiddleware` is enabled with `CSP.NONCE` in `SECURE_CSP` or `SECURE_CSP_REPORT_ONLY`, but `django.template.context_processors.csp` is not configured.

#### AI Assistance Disclosure (REQUIRED)

- [ ] **No AI tools were used** in preparing this PR.
- [x] **If AI tools were used**, I have disclosed which ones, and fully reviewed and verified their output.

AI assistance: ChatGPT/Codex helped prepare the patch. I reviewed and verified the changes and ran the targeted test suite locally.

#### Checklist

- [x] This PR follows the [contribution guidelines](https://docs.djangoproject.com/en/stable/internals/contributing/writing-code/submitting-patches/).
- [x] This PR **does not** disclose a security vulnerability (see [vulnerability reporting](https://docs.djangoproject.com/en/stable/internals/security/)).
- [x] This PR targets the `main` branch.
- [x] The commit message is written in past tense, mentions the ticket number, and ends with a period (see [guidelines](https://docs.djangoproject.com/en/dev/internals/contributing/committing-code/#committing-guidelines)).
- [x] I have not requested, and will not request, an automated AI review for this PR.
- [x] I have checked the "Has patch" ticket flag in the Trac system.
- [x] I have added or updated relevant tests.
- [x] I have added or updated relevant docs, including release notes if applicable.
- [ ] I have attached screenshots in both light and dark modes for any UI changes.

### CODE REVIEW

File: django/core/checks/security/base.py [hunk 1]
```diff
@@ -156,6 +170,36 @@ def _xframe_middleware():
...
+    return (
+        "django.middleware.csp.ContentSecurityPolicyMiddleware" in settings.MIDDLEWARE
+    )
+
+
+def _csp_policy_contains_nonce(policy):
+    if not isinstance(policy, dict):
+        return False
+    for values in policy.values():
+        if isinstance(values, list | tuple | set):
```
Comment (jacobtylerwalls) [inline django/core/checks/security/base.py]: Can you check `Iterable` instead? (Assuming you reject strings first.)

```py
In [1]: isinstance(frozenset(), set)
Out[1]: False
```

File: docs/releases/6.1.txt [hunk 2]
```diff
@@ -497,6 +497,11 @@ System checks
 * The :djadmin:`check` management command now supplies all ``databases`` if not
   specified. Callers should be prepared for databases to be accessed.
 
+* A new ``security.W027`` system check warns when
```
Comment (jacobtylerwalls) [inline docs/releases/6.1.txt]: I would move this to CSP; this section is more for the check framework itself.

Also, this goes under New Features > CSP, not backwards incompatible changes.

File: django/core/checks/security/base.py [hunk 3]
```diff
@@ -156,6 +170,36 @@ def _xframe_middleware():
...
+            return True
+    return False
+
+
+def _csp_context_processor_configured():
+    context_processor = "django.template.context_processors.csp"
+    django_templates = "django.template.backends.django.DjangoTemplates"
+    return any(
+        isinstance(template, dict)
+        and template.get("BACKEND") == django_templates
```
Comment (jacobtylerwalls) [inline django/core/checks/security/base.py]: Do you think we could loosen this assumption? For use in jinja, or eventually with https://github.com/LilyFirefly/django-rusty-templates?

```suggestion
```

cc/ @robhudson

**Review by jacobtylerwalls — COMMENTED** [review]
Thanks for the PR ⚡

**Comment by MiladZarour** [comment]
Thanks for the review @jacobtylerwalls.

I updated the PR to address your comments:
- Switched CSP.NONCE detection to check Iterable values while rejecting strings first, so frozenset and other iterable directive values are handled.
- Loosened the context processor check so it does not assume DjangoTemplates only; any template backend with django.template.context_processors.csp configured in OPTIONS["context_processors"] satisfies the check. Thanks also to @robhudson for the context here.
- Moved the release note under What's new > Minor features > CSP.

I also ran:
- python tests/runtests.py check_framework.test_security
- Ran 76 tests successfully, with 1 skipped.
- python -m black --check django/core/checks/security/base.py tests/check_framework/test_security.py
- python -m flake8 django/core/checks/security/base.py tests/check_framework/test_security.py

**Review by jacobtylerwalls — APPROVED** [review]
Welcome aboard ⛵ 

@robhudson, do you want to take a look?

File: docs/ref/checks.txt [hunk 4]
```diff
@@ -594,6 +594,9 @@ The following checks are run if you use the :option:`check --deploy` option:
   security-critical features will be vulnerable to attack.
 * **security.E026**: The CSP setting ``<SETTING_NAME>`` must be a dictionary
   (got ``<value>`` instead).
+* **security.W027**: Your CSP policy includes ``CSP.NONCE`` and
+  :class:`~django.middleware.csp.ContentSecurityPolicyMiddleware` is enabled,
+  but ``django.template.context_processors.csp`` is not configured.
```
Comment (robhudson) [inline docs/ref/checks.txt:595]: This is in the section:

> The following checks are run if you use the :option:`check --deploy` option:

But it's not a `deploy=True` check, so it should go down into the list below.

The same for `security.E026`, funnily enough.

**Review by robhudson — APPROVED** [review]
This looks great to me. I just noted a small change for the docs.

File: docs/ref/checks.txt (hunk 4 above)
Comment (jacobtylerwalls) [reply docs/ref/checks.txt:595]: Good catch, I'll fix E026 separately, and then I'll fix W027 here when resolving merge conflicts. Thanks for having a look!

**Comment by jacobtylerwalls** [comment]
I take it a warning is better than an error here because there could be other ways of providing the nonce to the request?

File: docs/ref/checks.txt (hunk 4 above)
Comment (jacobtylerwalls) [reply docs/ref/checks.txt:595]: Actually @MiladZarour, do you mind applying that feedback + fixing conflicts?

Comment (MiladZarour) [reply docs/ref/checks.txt:595]: Sure , will do that

File: django/core/checks/security/base.py [hunk 5]
```diff
@@ -145,6 +146,19 @@
     id="security.E026",
 )
 
+W027 = Warning(
+    "Your CSP policy includes CSP.NONCE and "
```
Comment (codingjoe) [inline django/core/checks/security/base.py]: The P in CSP means Policy. I know people also say SMS message… so I maybe just write `Content Security Policy` without the acronym. I'd generally try to avoid acronyms to improve readability for the uninitiated.
```suggestion
    "Your Content Security Policy includes CSP.NONCE and "
```

File: django/core/checks/security/base.py [hunk 6]
```diff
@@ -156,6 +170,36 @@ def _xframe_middleware():
     )
 
 
+def _csp_middleware():
+    return (
+        "django.middleware.csp.ContentSecurityPolicyMiddleware" in settings.MIDDLEWARE
+    )
```
Comment (codingjoe) [inline django/core/checks/security/base.py:176]: Since the function is only used once, I don't know if you save space or complexity wrapping the boolean statement.

File: django/core/checks/security/base.py [hunk 7]
```diff
@@ -156,6 +170,36 @@ def _xframe_middleware():
...
+
+
+def _csp_policy_contains_nonce(policy):
+    if not isinstance(policy, dict):
+        return False
```
Comment (codingjoe) [inline django/core/checks/security/base.py:183]: When would this case happen?

File: django/core/checks/security/base.py (hunk 1 above)
Comment (codingjoe) [inline django/core/checks/security/base.py]: EAFP: You might want to force an exception and handle it instead of type checking for all iterables (which you can't in Python anyway, since every object can implement `__iter__`).
```python
return CSP.NONCE in values
```

File: django/core/checks/security/base.py [hunk 8]
```diff
@@ -156,6 +170,36 @@ def _xframe_middleware():
...
+                return True
+        elif values == CSP.NONCE:
+            return True
+    return False
+
+
+def _csp_context_processor_configured():
+    context_processor = "django.template.context_processors.csp"
+    django_templates = "django.template.backends.django.DjangoTemplates"
+    return any(
```
Comment (codingjoe) [inline django/core/checks/security/base.py:195]: Shouldn't `all` backends include the processor?

**Review by codingjoe — COMMENTED** [review]
Hi there,

Since I have been working on other CSP topics, I figured you'd appreciate the review. Sadly, GitHub was down, so my comments are a bit older, but they should be complete.

Cheers!
Joe

**Comment by MiladZarour** [comment]
Hi @codingjoe , @jacobtylerwalls and @robhudson 

I will check the reviewed comments above and create a new push today later

File: django/core/checks/security/base.py (hunk 6 above)
Comment (jacobtylerwalls) [reply django/core/checks/security/base.py:176]: (I had this thought as well, but noticed the author was blending in with the existing pattern in the file.)

File: django/core/checks/security/base.py (hunk 7 above)
Comment (jacobtylerwalls) [reply django/core/checks/security/base.py:183]: We have a separate system check for this (security.E026), so I'm inclined to say handling it gracefully is a good idea?

File: django/core/checks/security/base.py (hunk 8 above)
Comment (jacobtylerwalls) [reply django/core/checks/security/base.py:195]: What about a template backend for emails?

File: django/core/checks/security/base.py (hunk 7 above)
Comment (codingjoe) [reply django/core/checks/security/base.py:183]: Based on my other comment, I would suggest something like:

```python
try:
    return CSP.NONCE in values
except TypeError:
    return False
```

**Comment by MiladZarour** [comment]
Thanks @codingjoe and @jacobtylerwalls.

That makes sense. I’ll update the warning text to say “Content Security Policy” and adjust the nonce check toward the EAFP style while still avoiding string false positives.

For the template backend check, I’ll keep it as “at least one backend” rather than requiring all backends, since a project may have separate backends for emails or other templates that don’t render request HTML.

I’ll also move security.W027 to the non-deploy checks section in docs/ref/checks.txt and resolve the conflicts in docs/ref/checks.txt and docs/releases/6.1.txt.

**Comment by MiladZarour** [comment]
Updated again to address the latest review comments:
- Changed the warning wording to “Content Security Policy”.
- Updated nonce detection to use EAFP-style membership checking while still avoiding string false positives.
- Moved security.W027 to the non-deploy security checks section in docs/ref/checks.txt.
- Preserved the latest CSP release note changes from main while resolving the docs/releases/6.1.txt conflict.

I also ran:
- python -m black --check django/core/checks/security/base.py tests/check_framework/test_security.py
- python -m flake8 django/core/checks/security/base.py tests/check_framework/test_security.py
- python tests/runtests.py check_framework.test_security
- Ran 76 tests successfully

**Comment by jacobtylerwalls** [comment]
(Looks like there are still conflicts to fix.)

File: django/core/checks/security/base.py (hunk 7 above)
Comment (jacobtylerwalls) [reply django/core/checks/security/base.py:183]: @MiladZarour I think the suggestion here was to replace the isinstance(..., dict) with except.

Comment (MiladZarour) [reply django/core/checks/security/base.py:183]: Thanks, got it. I’ll replace the top-level isinstance(policy, dict) guard with an EAFP check around policy.values(), while keeping the string guard for directive values.

File: django/core/checks/security/base.py (hunk 8 above)
Comment (codingjoe) [reply django/core/checks/security/base.py:195]: Rendering HTML for HTTP with two engines is a stretch, but you can silence the warning… and it's not an error.

I think the argument can be made both ways… I honestly don't care either way.

**Comment by MiladZarour** [comment]
Thanks, updated. I replaced the top-level isinstance(policy, dict) guard with an EAFP check around policy.values(), while keeping the string guard for directive values.

File: django/core/checks/security/base.py (hunk 7 above)
Comment (codingjoe) [reply django/core/checks/security/base.py:183]: I am an EAFP fanboy. But CPU branching isn't really a concern here. So 🤷‍♂️ 6-7

File: django/core/checks/security/base.py [hunk 9]
```diff
@@ -156,6 +170,40 @@ def _xframe_middleware():
...
+    except AttributeError:
+        return False
+    for values in policy_values:
+        if values == CSP.NONCE:
+            return True
+        if isinstance(values, str):
+            continue
+        try:
+            if CSP.NONCE in values:
+                return True
```
Comment (codingjoe) [inline django/core/checks/security/base.py]:
```suggestion
        return values == CSP.NONCE or CSP.NONCE in values
```
Same logic, right?

Comment (codingjoe) [reply django/core/checks/security/base.py]: It would also fail for a string, since the nonce object would actually be inside a string typed value.

**Review by codingjoe — COMMENTED** [review]
Wonderful!

I still believe we can ever so slightly shorten the implementation, but the logic looks correct to me.

File: django/core/checks/security/base.py (hunk 8 above)
Comment (jacobtylerwalls) [reply django/core/checks/security/base.py:195]: I think if we leave it as `any` and slightly favor false negatives over false positives, we'll still be alerting folks to the basic misconfiguration, "forgot about the context processor".

File: django/core/checks/security/base.py (hunk 6 above)
Comment (MiladZarour) [reply django/core/checks/security/base.py:176]: Thanks. I’ll keep _csp_middleware() since it follows the existing helper pattern in this file, as Jacob noted.

File: django/core/checks/security/base.py (hunk 7 above)
Comment (MiladZarour) [reply django/core/checks/security/base.py:183]: Updated in aab13cc. _csp_policy_contains_nonce() now uses an EAFP check around policy.values() instead of isinstance(policy, dict), so invalid CSP setting types are still handled gracefully while security.E026 reports them separately.

File: django/core/checks/security/base.py (hunk 8 above)
Comment (MiladZarour) [reply django/core/checks/security/base.py:195]: Agreed. I’ll keep this as “any backend” so the check favors avoiding false positives for projects with separate template backends, while still catching the common “forgot the context processor” misconfiguration.

File: django/core/checks/security/base.py (hunk 9 above)
Comment (MiladZarour) [reply django/core/checks/security/base.py]: Thanks, yes, the scalar and membership checks can be combined. I’ll simplify this while keeping the loop over all policy values and the TypeError handling, so later directives are still checked if an earlier one doesn’t contain CSP.NONCE.

**Comment by MiladZarour** [comment]
simplified in the latest push. I combined the scalar and membership checks while keeping the loop over all policy values and the TypeError handling.

**Review by codingjoe — APPROVED** [review]
Great, looks good to me!

File: django/core/checks/security/base.py (hunk 9 above)
Comment (codingjoe) [reply django/core/checks/security/base.py]: Good point!

File: django/core/checks/security/base.py (hunk 7 above)
Comment (jacobtylerwalls) [reply django/core/checks/security/base.py:183]: It occurs to me now that we could add test coverage for this by testing `SECURE_CSP=None`.

See: "# CSP settings must be a dictionary or None." in `check_csp_settings` below.

Comment (MiladZarour) [reply django/core/checks/security/base.py:183]: Good point. I’ll add coverage for SECURE_CSP=None / SECURE_CSP_REPORT_ONLY=None to confirm the nonce context processor check handles allowed None values without warning or crashing.

**Comment by MiladZarour** [comment]
added coverage for SECURE_CSP=None / SECURE_CSP_REPORT_ONLY=None to confirm the nonce context processor check handles allowed None values without warning or crashing.

I reran the targeted check_framework.test_security tests: 76 passed with 1 skipped.

**Comment by MiladZarour** [comment]
Added coverage for a scalar CSP directive value to exercise the TypeError path without warning.

**Review by jacobtylerwalls — APPROVED** [review]
